src/backend/distributed/commands/utility_hook.c

```diff
@@ -344,26 +344,52 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 			}
 			ResetConstraintDropped();
 
+			/*
+			 * We're only interested in top-level CREATE TABLE commands
+			 * to create a tenant table or a Citus managed table.
+			 */
+			Oid createdRelationId = InvalidOid;
 			if (context == PROCESS_UTILITY_TOPLEVEL &&
-				ShouldAddNewTableToMetadata(parsetree))
+				(IsA(parsetree, CreateStmt) ||
+				 IsA(parsetree, CreateForeignTableStmt)))
 			{
+				CreateStmt *baseCreateTableStmt =
+					IsA(parsetree, CreateStmt) ? (CreateStmt *) parsetree :
+					(CreateStmt *) &(((CreateForeignTableStmt *) parsetree)->base);
+
 				/*
 				 * Here we need to increment command counter so that next command
 				 * can see the new table.
 				 */
 				CommandCounterIncrement();
-				CreateStmt *createTableStmt = (CreateStmt *) parsetree;
-				Oid relationId = RangeVarGetRelid(createTableStmt->relation,
-												  NoLock, false);
+				bool missingOk = false;
+				createdRelationId = RangeVarGetRelid(baseCreateTableStmt->relation,
+													 NoLock, missingOk);
+			}
 
+			/*
+			 * Check ShouldCreateTenantSchemaTable() before ShouldAddNewTableToMetadata()
+			 * because we don't want to unnecessarily add the table into metadata
+			 * (as a Citus managed table) before distributing it as a tenant table.
+			 */
+			if (!OidIsValid(createdRelationId))
+			{
+				/* no tables were created by this command, or it wasn't a top-level one */
+			}
+			else if (ShouldCreateTenantSchemaTable(createdRelationId))
+			{
+				CreateTenantSchemaTable(createdRelationId);
+			}
+			else if (ShouldAddNewTableToMetadata(createdRelationId))
+			{
 				/*
 				 * Here we set autoConverted to false, since the user explicitly
 				 * wants these tables to be added to metadata, by setting the
 				 * GUC use_citus_managed_tables to true.
 				 */
 				bool autoConverted = false;
 				bool cascade = true;
-				CreateCitusLocalTable(relationId, cascade, autoConverted);
+				CreateCitusLocalTable(createdRelationId, cascade, autoConverted);
 			}
 		}
```

Comment:
Can we move this logic into a function if possible? To make `multi_processutility` function, which is already long enough, more readable.

Reply:
For such slightly complex logics, I feel like doing the things explicitly is better than hiding the details. But ofc I'm open to suggestions too, any ideas about how to improve there?

Reply:
seems like we could do something like:
```c
if (context == PROCESS_UTILITY_TOPLEVEL &&
	(IsA(parsetree, CreateStmt) || IsA(parsetree, CreateForeignTableStmt)))
{
    ConvertNewTableIfNecessary(parsetree);
}
```